pauliopt/gates.py

```diff
@@ -112,18 +116,158 @@ def gadgets(self):
     def to_qiskit(self):
         pass
 
+    @abstractmethod
+    def inverse(self):
+        pass
+
+    def apply_permutation(self, permutation: list) -> None:
+        register = list(range(len(permutation)))
+        self.qubits = tuple(
+            [permutation[register.index(qubit)] for qubit in self.qubits]
+        )
 
-class PhaseGate(Gate):
-    def __init__(self, phase, *qubits):
+    def copy(self):
+        return self.__class__(*self.qubits)
+
+
+class PhaseGate(Gate, ABC):
+    def __init__(self, phase: Angle, *qubits):
         super().__init__(*qubits)
         self.phase = phase
 
     def __repr__(self) -> str:
         args = map(repr, self.qubits)
         return f"{self.name}({self.phase}, {', '.join(args)})"
 
+    def get_phase_as_float(self):
+        return self.phase if not isinstance(self.phase, Angle) else self.phase.to_qiskit
```

Comment:
so weird that this is `to_qiskit`, does `float(.)` not work?

Reply:
That is the implementation of `to_qiskit`:

```
    @property
    def to_qiskit(self) -> float:
        return float(self)
```

The whole angle type object is kind of a difficult and un-understandable interface, might be sensible to also refactor it or remove the type completely.